### ISSUE TITLE
How does SFTTrainer handle instruction formatted datasets when a tokenizer has no chat_template?

### ISSUE THREAD

**Issue by JohnGiorgi** [huggingface/trl#1233]
Hi! I am interested in using the `SFTTrainer` for instruction-tuning. Following [the docs](https://huggingface.co/docs/trl/main/en/sft_trainer#dataset-format-support), I can see that I can provided examples in the following format to have the trainer format things for me:

```json
{"prompt": "<prompt text>", "completion": "<ideal generated text>"}
{"prompt": "<prompt text>", "completion": "<ideal generated text>"}
{"prompt": "<prompt text>", "completion": "<ideal generated text>"}
```

The docs also say:

> The [SFTTrainer](https://huggingface.co/docs/trl/main/en/trainer#trl.SFTTrainer) will then format the dataset for you using the defined format from the model’s tokenizer with the [apply_chat_template](https://huggingface.co/docs/transformers/main/en/chat_templating#templates-for-chat-models) method.

My question and confusion is, what does the trainer do if the tokenizer has no `chat_template`, as is the case with the [base llama model](https://huggingface.co/meta-llama/Llama-2-13b-hf/blob/main/tokenizer_config.json)?

**Comment by younesbelkada** [comment]
Hi @JohnGiorgi 
Thanks for the issue! Per my understanding if there is no chat template that feature is simply not supported - correct @philschmid ?

**Comment by younesbelkada** [comment]
TLDR is that if you need to use a chat dataset you need to use a model that supports chat templating - if you want to use that model I think you can clone it and manually add a chat template in that model

**Comment by philschmid** [comment]
Hey @JohnGiorgi,

There are fallback templates in `tranformers`. If the loaded model/tokenizer is not having a `chat_template` then transformers fallback to the [class-specific template](https://github.com/huggingface/transformers/blob/d93ef7d7512e79612606f29e6ae308920f0a86cd/src/transformers/models/llama/tokenization_llama.py#L415), if there is no class-specific template it falls back to [base chat template](https://github.com/huggingface/transformers/blob/d93ef7d7512e79612606f29e6ae308920f0a86cd/src/transformers/tokenization_utils_base.py#L1783), which is the `chatml` format.

**Comment by JohnGiorgi** [comment]
Thanks all! I understand now better and agree that it is probably best to explicitly define a template.

**Comment by philschmid** [comment]
#1242 might be interesting for you.
